File: statsig-pyo3/py_src/statsig_python_core/persistent_storage.py
from typing import Any, Dict, List, Optional
from typing_extensions import TypeAlias, TypedDict
from statsig_python_core import StatsigUser, PersistentStorageBaseClass

class StickyValues(TypedDict):
    value: bool
    json_value: Optional[Any]
    rule_id: str
    group_name: Optional[str]
    secondary_exposures: List[Dict[str, str]]
    explicit_parameters: Optional[List[str]]
    config_delegate: Optional[str]
    undelegated_secondary_exposures: List[Dict[str, str]]
    config_version: Optional[int]
    time: int

UserPersistedValues: TypeAlias = Dict[str, StickyValues]
PersistedValues: TypeAlias = Dict[str, UserPersistedValues]

class PersistentStorage(PersistentStorageBaseClass):
    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, *args, **kwargs):
        super().__init__()
        self.load_fn = self.load
        self.save_fn = self.save
        self.delete_fn = self.delete

    def load(self, key: str) -> Optional[UserPersistedValues]:
        pass

    def save(self, key: str, config_name: str, data: StickyValues):
        pass

    def delete(self, key: str, config_name: str):
        pass
    

    def get_user_persisted_value(self, user: StatsigUser, id_type: str) -> Optional[UserPersistedValues]:
        storage_key = self.get_storage_key(user, id_type) 
        if storage_key is not None:
            return self.load(storage_key)

        return None
    
    @staticmethod
    def get_storage_key(user: StatsigUser, id_type: str) -> Optional[str]:
        lower_case_id_type = id_type.lower()
        if (lower_case_id_type == "user_id" or lower_case_id_type == "userid"):
            id = getattr(user,"user_id")
            return f"{id}:userID"
        else:
            if (user.custom_ids is not None):
                id = user.custom_ids.get(id_type)
                return f"{id}:{id_type}"
        return None

File: statsig-go/persistent_storage.go
package statsig_go_core

import (
	"encoding/json"
	"fmt"
	"runtime"

	"github.com/statsig-io/statsig-go-core/internal"
)

type PersistentStorageFunctions struct {
	Load   func(key string) *UserPersistedValues
	Save   func(key string, configName string, data StickyValues)
	Delete func(key string, configName string)
}

type PersistentStorage struct {
	functions PersistentStorageFunctions
	ref       uint64
}

type SecondaryExposure struct {
	Gate      string `json:"gate"`
	GateValue string `json:"gateValue"`
	RuleID    string `json:"ruleID"`
}

type StickyValues struct {
	Value                         bool                `json:"value"`
	JSONValue                     map[string]string   `json:"json_value"`
	RuleID                        string              `json:"rule_id"`
	GroupName                     *string             `json:"group_name"`
	SecondaryExposures            []SecondaryExposure `json:"secondary_exposures"`
	UndelegatedSecondaryExposures []SecondaryExposure `json:"undelegated_secondary_exposures"`
	ConfigDelegate                *string             `json:"config_delegate"`
	ExplicitParameters            *[]string           `json:"explicit_parameters"`
	Time                          int64               `json:"time"`
	ConfigVersion                 *int64              `json:"config_version,omitempty"`
}

type UserPersistedValues map[string]StickyValues

type persistentStorageArgs struct {
	Key        string        `json:"key"`
	ConfigName string        `json:"config_name"`
	Data       *StickyValues `json:"data,omitempty"`
}

func NewPersistentStorage(functions PersistentStorageFunctions) *PersistentStorage {
	storage := &PersistentStorage{
		functions: functions,
		ref:       0,
	}

	storage.ref = GetFFI().persistent_storage_create(
		"go",
		// Load
		func(argsPtr *byte, argsLength uint64) *byte {
			keyStr := internal.GoStringFromPointer(argsPtr, argsLength)
			if keyStr == nil {
				return nil
			}

			result := storage.functions.Load(*keyStr)

			if result == nil {
				return nil
			}

			json, err := json.Marshal(*result)
			if err != nil {
				fmt.Println("Error marshalling user persisted values", err)
				return nil
			}

			return &json[0]
		},
		// Save
		func(argsPtr *byte, argsLength uint64) {
			data, err := tryMarshalPersistentStorageArgs(argsPtr, argsLength)
			if err != nil {
				fmt.Println("Error marshalling persistent storage args", err)
				return
			}

			if data.Data == nil {
				fmt.Println("Error marshalling persistent storage args: Data is nil")
				return
			}

			storage.functions.Save(data.Key, data.ConfigName, *data.Data)
		},
		// Delete
		func(argsPtr *byte, argsLength uint64) {
			data, err := tryMarshalPersistentStorageArgs(argsPtr, argsLength)
			if err != nil {
				fmt.Println("Error marshalling persistent storage args", err)
				return
			}
			storage.functions.Delete(data.Key, data.ConfigName)
		},
	)

	runtime.SetFinalizer(storage, func(obj *PersistentStorage) {
		GetFFI().persistent_storage_release(obj.ref)
	})

	return storage
}

func (c *PersistentStorage) INTERNAL_testPersistentStorage(action string, key string, configName string, data string) string {
	return GetFFI().__internal__test_persistent_storage(c.ref, action, key, configName, data)
}

func tryMarshalPersistentStorageArgs(inputPtr *byte, inputLength uint64) (*persistentStorageArgs, error) {
	data := internal.GoStringFromPointer(inputPtr, inputLength)

	var args persistentStorageArgs
	err := json.Unmarshal([]byte(*data), &args)
	if err != nil {
		fmt.Println("Error unmarshalling persistent storage args", err)
		return nil, err
	}

	return &args, nil
}
